test/test_torch.py

```diff
@@ -12840,6 +12769,208 @@ def test_function_unwrap_message(self):
+            ("sigmoid", doubles, False, False, 'cuda'),
+            ("sqrt", doubles, True, True, 'cpu'),
+            ("sqrt", doubles, False, True, 'cuda'),
+            ("tan", doubles, True, True, 'cpu'),
+            ("tan", doubles, False, True, 'cuda'),
+            ("tanh", doubles, True, True, 'cpu'),
+            ("tanh", doubles, False, True, 'cuda'),
+            ("trunc", doubles, True, True, 'cpu'),
+            ("trunc", doubles, False, True, 'cuda')
+        ]
+
+        for (fn, inputs, has_input_output_mem_overlap_check,
+                has_internal_mem_overlap_check, dev) in unary_mem_overlap_cases:
+            if dev != device:
+                continue
+            out_fn = getattr(torch, fn)
+            in_fn = getattr(torch.Tensor, fn + '_')
+
+            if has_input_output_mem_overlap_check:
+                self.unary_check_input_output_mem_overlap(inputs, sz, out_fn)
+            else:
+                with self.assertRaises(AssertionError):
+                    self.unary_check_input_output_mem_overlap(inputs, sz, out_fn)
+
+            if has_internal_mem_overlap_check:
+                self.unary_check_internal_mem_overlap(in_fn, dev)
+            else:
+                with self.assertRaises(AssertionError):
+                    self.unary_check_internal_mem_overlap(in_fn, dev)
+
+    def binary_check_internal_mem_overlap(self, inplace_op, device):
+        if isinstance(inplace_op, str):
+            inplace_op = getattr(torch.Tensor, inplace_op)
+        tensor = torch.tensor(42.0, device=device).expand(3, 3)
+        other = torch.rand_like(tensor)
+        with self.assertRaisesRegex(RuntimeError, 'single memory location'):
+            inplace_op(tensor, other)
+
+    def binary_check_input_output_mem_overlap(self, op, device):
+        sz = 3
+        data = torch.randn(3 * sz, device=device)
+        other = torch.randn(sz, device=device)
+
+        self.unary_check_input_output_mem_overlap(
+            data, sz, lambda input, out: op(other, input, out=out))
+
+        self.unary_check_input_output_mem_overlap(
+            data, sz, lambda input, out: op(input, other, out=out))
+
+    @torchtest.for_all_device_types()
+    def test_binary_op_mem_overlap(self, device):
+        ops = [
+            "add",
+            "mul",
+            "sub",
+            "div",
+            "logical_xor"
+        ]
+
+        for fn in ops:
+            out_op = getattr(torch, fn)
+            inplace_op = getattr(torch.Tensor, fn + '_')
+            # internal overlap check for inplace_op
+            self.binary_check_internal_mem_overlap(inplace_op, device)
+            # input/output overlap check for out op
+            self.binary_check_input_output_mem_overlap(out_op, device)
+
+    def ternary_check_input_output_mem_overlap(self, op, device):
+        sz = 3
+        other = torch.randn(sz, device=device)
+
+        self.binary_check_input_output_mem_overlap(
```

Comment:
So we are testing, for each input, that if it has partial memory overlap with the output tensor, then the function errors out.

In that case we should call `self.unary_check_input_output_mem_overlap` three times, one for each input, instead of calling "binary_check_input_output_mem_overlap" twice. It's not clear to me that all of the cases are being tested right now

Reply:
This being said, it would also be interesting if we could deduplicate `ternary_check_input_output_mem_overlap` and `binary_check_input_output_mem_overlap`, though it's not obvious to me if that is a good idea right now or not.

Reply:
addressed